cgp/genome.py

```diff
@@ -267,6 +267,172 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        used_idx_list = []
+
+        while len(addable_nodes) > 0:
+            # pick one of the addable nodes
+            current_idx = rng.choice(list(addable_nodes))
+
+            # take this nodes dna sequence
+            node_dna = self._dna[
+                current_idx * length_per_region : (current_idx + 1) * length_per_region
+            ]
+            # and insert it at the working location of the copy
+            dna_for_reorder[
+                node_idx * length_per_region : (node_idx + 1) * length_per_region
+            ] = node_dna
+
+            # remove the current idx from the dependencies set
+            for current_dependencies in dependencies:
+                current_dependencies.discard(current_idx)
+
+            # update set of addable nodes
+            addable_nodes = set(
+                idx
+                for (idx, item) in enumerate(dependencies)
+                if not item and self._is_hidden_region(idx)
+            )
+            # remove current_idx and all previous used idx's
+            # from addable_nodes and increment node_idx counter
+            used_idx_list.append(current_idx)
+            addable_nodes = addable_nodes.difference(used_idx_list)
+            node_idx += 1
+
+        # change the gene values such that they fit the new location of the genes
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
+            dna_for_reorder[gene_idx] = gene_value
+
+        # change the invalid gene values (coming from unused arities)
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by smaller value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+        self.dna = dna_for_reorder
+
+    def _determine_dependencies(self) -> List[set]:
+
+        dependencies = []
+        for region_idx in range(self._n_regions):
+            current_node_dependencies = set()
+
+            if self._is_input_region(region_idx):
+                dependencies.append(
+                    current_node_dependencies
+                )  # input regions have no dependencies
+
+            elif self._is_hidden_region(
+                region_idx
+            ):  # can assume dependencies is not empty anymore
```

Comment:
because first empty sets for the input regions were added to `dependencies`. Anyway this is redundant now, if i change `dependencies` to a dict, as you suggested above